samples/subsys/tracing/src/gpio_main.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2024 Tenstorrent AI ULC
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+#include <zephyr/logging/log.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/drivers/gpio/gpio_emul.h>
+
+
+/* size of stack area used by each thread */
+#define STACKSIZE (2048)
+
+/* scheduling priority used by each thread */
+#define PRIORITY 7
+
+
+void test_handler(const struct device *port, struct gpio_callback *cb, uint32_t pins)
+{
+	printk("Interrupt detected!\n");
+}
+
+
+void gpio_sample(void)
+{
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(gpio0));
+
+	/* Configure pin 0 as output and toggle */
+	gpio_pin_configure(dev, 0, GPIO_OUTPUT);
+	gpio_pin_set(dev, 0, 1);
+	gpio_pin_set(dev, 0, 0);
+
+	/* Configure pin 1 as input */
+	gpio_pin_configure(dev, 1, GPIO_INPUT);
+
+	/* Read pin 1 */
+	if (DT_HAS_COMPAT_STATUS_OKAY(zephyr_gpio_emul)) {
```

Comment:
`DT_HAS_COMPAT_STATUS_OKAY()` checks if there's any `zephyr_gpio_emul` that has status "okay" in the devicetree, but not necessarily the one that you are using (`gpio0`).

`gpio0` must have status "okay", otherwise `DEVICE_DT_GET(DT_NODELABEL(gpio0))` fails to compile.

To check if `gpio0` has been initialized properly, you can do `device_is_ready(dev)` check after L29.

The repeated `if (DT_HAS_COMPAT_STATUS_OKAY(zephyr_gpio_emul))` checks are redundant after that.